src/kafka_endpoint.cc

```diff
@@ -0,0 +1,971 @@
+    : server_(server), shm_manager_(shm_manager), port_(port),
+      consumer_topics_(consumer_topics.begin(), consumer_topics.end()),
+      producer_topic_(producer_topic)
+
+{
+  allocator_ = nullptr;
+  FAIL_IF_ERR(
+      TRITONSERVER_ResponseAllocatorNew(
+          &allocator_, InferResponseAlloc, InferResponseFree,
+          nullptr /* start_fn */),
+      "creating response allocator");
+  FAIL_IF_ERR(
+      TRITONSERVER_ResponseAllocatorSetQueryFunction(
+          allocator_, OutputBufferQuery),
+      "setting allocator's query function");
+  FAIL_IF_ERR(
+      TRITONSERVER_ResponseAllocatorSetBufferAttributesFunction(
+          allocator_, OutputBufferAttributes),
+      "setting allocator's buffer attributes function");
+}
+
+KafkaEndpoint::~KafkaEndpoint()
+{
+  IGNORE_ERR(Stop());
+  LOG_TRITONSERVER_ERROR(
+      TRITONSERVER_ResponseAllocatorDelete(allocator_),
+      "deleting response allocator");
+}
+
+
+TRITONSERVER_Error*
+KafkaEndpoint::Create(
+    const std::shared_ptr<TRITONSERVER_Server>& server,
+    const std::shared_ptr<SharedMemoryManager>& shm_manager,
+    const std::string& port, const std::vector<std::string>& consumer_topics,
+    const std::string& producer_topic,
+    std::unique_ptr<KafkaEndpoint>* kafka_endpoint)
+{
+  if (port.empty() || consumer_topics.empty() || producer_topic.empty()) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INVALID_ARG,
+        "Kakfa [port], [producer topic], or [consumer topics] not specified");
+  }
+  kafka_endpoint->reset(new KafkaEndpoint(
+      server, shm_manager, port, consumer_topics, producer_topic));
+
+  LOG_INFO << "Started Kafka Endpoint. Specified broker port: " << port;
+
+  return nullptr;  // Success
+}
+
+TRITONSERVER_Error*
+KafkaEndpoint::Start()
+{
+  RETURN_IF_ERR(StartConsumer());
+  RETURN_IF_ERR(StartProducer());
+  return nullptr;  // Success
+}
+
+TRITONSERVER_Error*
+KafkaEndpoint::StartProducer()
+{
+  if (producer_ != nullptr) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_ALREADY_EXISTS,
+        "Kafka Endpoint producer is already running!");
+  }
+  kafka::Properties props({
+      {"bootstrap.servers", port_},
+      {"enable.idempotence", "true"},
+  });
+  // Create a producer instance.
+  producer_ = std::make_unique<kafka::clients::KafkaProducer>(props);
+
+  return nullptr;  // Success
+}
+
+TRITONSERVER_Error*
+KafkaEndpoint::StartConsumer()
+{
+  if (consumer_thread_.joinable()) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_ALREADY_EXISTS,
+        "Kafka Endpoint consumer is already running!");
+  }
+  try {
+    kafka::Properties props(
+        {{"bootstrap.servers", port_}, {"enable.auto.commit", "true"}});
+
+    // Create a consumer instance
+    consumer_ = std::make_unique<kafka::clients::KafkaConsumer>(props);
+
+    // Subscribe to topics
+    // TODO: Currently if a single topic in a list is valid, the server
+    // will not time out, it will silently fail.
```

Comment:
Can you elaborate on this gap a little bit here? Also reference a DLIS-XXX ticket for followup. The ticket can provide more details on the current vs expected behavior. My suggestion below might not be accurate so please correct as needed.
```suggestion
    // TODO: If there is a mix of valid and invalid consumer topics here, the server
    // will not timeout due to the valid topics, but will silently fail for the invalid topics.
```